Fixes to belongsToMany relations and correct hasOne relation + customizations

Alamantus opened the pull request:
I took PR #7 and made the changes I suggested in my review of that PR, plus I added the ability to customize the line color and arrow shapes for each relationship type. I also updated the README to explain these customization options.

Alamantus commented:
Sorry, I just realized I could add more customization options, so I did!

zekenie commented on index.js:
```diff
@@ -2,14 +2,80 @@ const Vis = require("./graphvis");
+  const mappings = [];
+
+  const mapper = {
+    associationTypes: new Map([
```
Why are we using a `Map` here?

zekenie replied:
Could we replace this section with

```
const associationTypes = {
  ...defalultAssociationTypes,
  ...arrowShapes
}
```

?

Alamantus replied:
I feel like we probably could do that! I'll take a look when I'm able and report back.

Alamantus replied:
In the most recent commit, I simplified this to not use `Map`

zekenie commented on index.js:
```diff
@@ -2,14 +2,80 @@ const Vis = require("./graphvis");
+        typeArrowShapes: arrowShapes['BelongsToMany']
+          ? arrowShapes['BelongsToMany']
+          : ['none', 'crow'],
+        items: []
```
I have a preference for storing the items in a different object. I feel like one object to store the rendering configuration, and another to store the model associations will be easier to understand

Alamantus replied:
I think that makes sense, especially since a later step just ends up rewriting the associations anyway. I think the idea for a separate object is the piece I was missing to simplify the deduplication process!

Alamantus replied:
I simplified this down to separate objects that handle a little better.

zekenie commented on index.js:
```diff
@@ -2,14 +2,80 @@ const Vis = require("./graphvis");
+  };
+
+  // Separation of association by type to process in the right order
+  associations.forEach(association =>
```
What if we consider something like

```
const associationsByType = _.groupBy(associations, association => association.associationType)
```

Alamantus replied:
Is that lodash? That would probably work great, but it would add a dependency. I don't think there's a direct equivalent function in vanilla JS, is there?

zekenie replied:
You're correct, I don't think there is. But I think it's very expressive, and I support adding the dependency

Alamantus replied:
Ok, what do you think of this option instead of installing lodash:

```
const groupedAssociations = associations.reduce((result, association) => {
  if (typeof result[association.associationType] === 'undefined')
    result[association.associationType].push(association);
  else
    result[association.associationType] = [association];
}, {});
```

It accomplishes the same thing in just a few extra rows without extra dependencies.

If you'd rather add lodash, I can do that, too, but I feel that fewer dependencies is typically better.

Alamantus replied:
I used the above solution (except I fixed the glaring error of not returning `result` after modifying it :sweat_smile:) and it's pretty good!

zekenie commented on index.js:
```diff
@@ -2,14 +2,80 @@ const Vis = require("./graphvis");
+    .map(
+      entries =>
+      entries.length == 1
+      ? `"${entries[0][0]}" -> "${entries[0][0]}" [arrowtail=${entries[0][1]}, arrowhead=none, dir=both, arrowsize=${arrowSize.toString()}]`
```
I'm having a hard time following what's going on here. Could we do anything to simplify, or maybe add a comment if it can't be simplified

Alamantus replied:
This is probably the most confusing part of the code from PR #7... I tried to quickly modify it to make it a little bit less confusing, but it'll take either a rewrite that'll make the resulting code longer or I can try to add a comment to explain what's happening.

Alamantus replied:
I figured this out, and this implementation is VERY clever, so I kept it in and added comments instead! It saves a step, and with the comment, it's understandable, so I'd recommend we keep it!

Alamantus commented:
I updated the `relationships()` function to be easier to follow, and it reduced the code size as a result! I also added some extra comments and reduced the number of variables used.

@zekenie Let me know if there's anything else I should change!

zekenie commented on index.js:
```diff
@@ -2,14 +2,65 @@ const Vis = require("./graphvis");
+    HasOne: ['none', 'none'],
+  }, customArrowShapes);
+
+  const groupedAssociations = associations.reduce((result, association) => {
```
For readability I'd like to use lodash's group by

Alamantus replied:
I've used lodash's `groupBy()` in the next commit!

zekenie commented on index.js:
```diff
@@ -2,14 +2,65 @@ const Vis = require("./graphvis");
+    'HasOne',
+  ].reduce((mapResult, type) => {
+    (
+      typeof groupedAssociations[type] === 'undefined'
```
Stylistic preference here, but I'd like to see this expressed as

```js
const associtionsOfType = groupedAssociations[type] || [];
for (const association of associationsOfType) {
  ...
}
```

Alamantus replied:
I've made this change in the next commit. The `||` syntax didn't come to mind when I first did it, and I like it better, too.

zekenie commented on index.js:
```diff
@@ -2,14 +2,65 @@ const Vis = require("./graphvis");
+    return result;
+  }, {});
+
+  const mappings = [  // Follow this heirarchy to ensure the arrows are set up correctly, eg. HasOne specifications override BelongsTo
```
What does mappings really mean?

Alamantus replied:
Lol good question, but I'm not sure what variable name would work better. I'm just using @wwmoraes' naming here since I forked from him.

Alamantus replied:
I came up with some new variable names for this whole section in the next commit that are hopefully a little bit more descriptive.

zekenie commented on index.js:
```diff
@@ -37,7 +88,7 @@ const modelTemplate = ({ model, columns }) => `"${
     ${
       columns
-        ? Object.values(model.attributes)
+        ? Object.values(model.rawAttributes)
```
is there a reason for this change? What does it do?

Alamantus replied:
This is in response to issue #10, solved by [this comment](https://github.com/zekenie/sequelize-erd/issues/10#issuecomment-525941258). Apparently Sequelize 5 changed something so the values targeted here are accessible via `rawAttributes` instead of `attributes`.

Alamantus commented:
I addressed the style things and left a comment on `rawAttributes` in your review. :)
